samples: matter: Introduce Timed Schedule Access in Lock (#15485)
nrfconnect/sdk-nrf: samples/matter/lock/README.rst

         * ``SetYearDaySchedule``
         * ``GetYearDaySchedule``
         * ``GetYearDayScheduleResponse``
         * ``ClearYearDaySchedule``

      * For the ``holiday`` schedule support:

         * ``SetHolidaySchedule``
         * ``GetHolidaySchedule``
         * ``GetHolidayScheduleResponse``
         * ``ClearHolidaySchedule``

#. Save the lock.zap file, and close ZAP-tool.
#. Generate the new ZAP files according to changes in the Door Lock cluster by utilizing the following west command:

   .. code-block:: console

         west zap-generate

#. Enable the feature in |NCS| Matter Lock sample by setting the :kconfig:option:`CONFIG_LOCK_SCHEDULES` kconfig option to ``y``.
#. Use the following kconfig options to modify maximum number of the specific schedule type:

   - :kconfig:option:`CONFIG_LOCK_MAX_WEEKDAY_SCHEDULES_PER_USER` to define the maximum number of week-day schedules per one user.
   - :kconfig:option:`CONFIG_LOCK_MAX_YEARDAY_SCHEDULES_PER_USER` to define the maximum number of year-day schedules per one user.
   - :kconfig:option:`CONFIG_LOCK_MAX_HOLIDAY_SCHEDULES` to define the maximum number of holiday schedules.

> To learn more how to configure Door Lock cluster see the :ref:`ug_matter_creating_accessory` user guide.

Reviewer: ```suggestion
To learn more about configuring the Door Lock cluster, see the :ref:`ug_matter_creating_accessory` user guide.
```

> All schedule timed-access entries are saved to non-volatile memory and loaded automatically after device reboot.

Reviewer: ```suggestion
All scheduled timed access entries are saved to non-volatile memory and loaded automatically after device reboot.
```

> To disable the feature you need to revert all changes in the :file:`lock.zap` file, re-generate the ZAP files and set the :kconfig:option:`CONFIG_LOCK_SCHEDULES` kconfig option to ``n``.

Reviewer: ```suggestion
To disable the feature, you need to revert all changes in the :file:`lock.zap` file, re-generate the ZAP files and set the :kconfig:option:`CONFIG_LOCK_SCHEDULES` Kconfig option to ``n``.
```

      * ``NumberOfHolidaySchedulesSupported`` for the ``Holiday`` schedule support.

   #. In the **Door Lock** context window, go to the **Attribute Reporting** page and enable all required attribute reporting entries:

      * ``NumberOfWeekDaySchedulesSupportedPerUser`` for the ``Week-day`` schedule support.
      * ``NumberOfYearDaySchedulesSupportedPerUser`` for the ``Year-day`` schedule support.
      * ``NumberOfHolidaySchedulesSupported`` for the ``Holiday`` schedule support.

   #. In the **Door Lock** context window, go to the **Commands** page and enable all required command entries:

      * For the ``Week-day`` schedule support:

        * ``SetWeekDaySchedule``
        * ``GetWeekDaySchedule``
        * ``GetWeekDayScheduleResponse``
        * ``ClearWeekDaySchedule``

      * For the ``Year-day`` schedule support:

         * ``SetYearDaySchedule``
         * ``GetYearDaySchedule``
         * ``GetYearDayScheduleResponse``
         * ``ClearYearDaySchedule``

      * For the ``Holiday`` schedule support:

         * ``SetHolidaySchedule``
         * ``GetHolidaySchedule``
         * ``GetHolidayScheduleResponse``

>          * ``ClearHolidaySchedule``
> #. In the **Door Lock** context window, go to the **Attributes** tab and set the proper bits for the ``FeatureMap`` attribute:

Reviewer: ```suggestion
         * ``ClearHolidaySchedule``

#. In the **Door Lock** context window, go to the **Attributes** tab and set the proper bits for the ``FeatureMap`` attribute:
```

   #. In ZAP Tool GUI select the endpoint 0 and enable Time Synchronization cluster, with both server and client roles, for that endpoint.

   #. Click the :guilabel:`Configure` symbol for the ``Time Synchronization`` cluster entry.
   #. In the **Time Synchronization** context window, go to the **Attributes** tab and enable all required attributes:

      * ``UTCTime``.
      * ``Granularity``.
      * ``TimeZone``.
      * ``DSTOffset``.
      * ``LocalTime``.
      * ``TimeZoneDatabase``.
      * ``TimeZoneListMaxSize``.
      * ``DSTOffsetListMaxSize``.
      * ``TrustedTimeSource``.

   #. In the **Time Synchronization** context window, go to the **Attribute Reporting** page and enable all available attribute reporting entries.

   #. In the **Time Synchronization** context window, go to the **Commands** page and enable all required command entries:

      * ``SetUTCTime``.
      * ``SetTimeZone``.
      * ``SetTimeZoneResponse``.
      * ``SetDSTOffset``.
      * ``SetTrustedTimeSource``.

#. In the **Time Synchronization** context window, go to the **Attributes** tab and set the proper bits for the ``FeatureMap`` attribute:

>    * For the ``TimeZone`` support set the ``0th`` bit of the feature map bit map.

Reviewer: ```suggestion
   * For the ``TimeZone`` support, set the ``0th`` bit of the feature map bit map.
```

   * ``DaysMask`` is a bitmap of numbers of the week days starting from ``0`` as a Sunday and finishing at ``6`` as a Saturday.
     For example, to assign this schedule to Tuesday, Thursday and Saturday you need to provide ``84`` because it is equivalent to the ``01010100`` bitmap.
   * ``StartHour`` is the starting hour for the week day schedule.
   * ``StartMinute`` is the starting minute for the week day schedule.
   * ``EndHour`` is the ending hour for the week day schedule.
   * ``EndMinute`` is the ending hour for the week day schedule.
   * ``node-id`` is the device node ID.
   * ``endpoint-id`` is the Matter door lock endpoint, in this sample assigned to ``1``.

   For example, use the following command to set a ``week-day`` schedule for Tuesday, Thursday and Saturday to start at 7:30 AM and finish at 10:30 AM dedicated for user with ID ``1``:

   .. code-block:: console

      ./chip-tool doorlock set-week-day-schedule 1 1 84 7 30 10 30 1 1

#. Set the example ``year-day`` schedule using the following command:

   ./chip-tool doorlock set-year-day-schedule YearDayIndex UserIndex LocalStartTime LocalEndTime node-id endpoint-id

   Where:

>    * ``YearDayIndex`` is an index of the new schedule, starting from ``1``.

Reviewer: ```suggestion
   * *yearday-index* is the index of the new schedule, starting from ``1``.
```

     The maximum value is defined by the :kconfig:option:`CONFIG_LOCK_MAX_YEARDAY_SCHEDULES_PER_USER` kconfig option.
   * ``UserIndex`` is the user index defined in the previous step, starting from ``1``.
   * ``LocalStartTime`` is the starting time in Epoch Time.
   * ``LocalEndTime`` is the ending time in Epoch Time.
   * ``node-id`` is the device node ID.
   * ``endpoint-id`` is the Matter door lock endpoint, in this sample assigned to ``1``.

   Both ``LocalStartTime`` and ``LocalEndTime`` are in seconds with the local time offset based on the local timezone and DST offset on the day represented by the value.

>    For example:

Reviewer: I'd expand this to a full sentence that explains the values used, like was done for `Week-day`.

Author: Done

   .. code-block:: console

>       ./chip-tool doorlock set-week-day-schedule 1 1 60 120 1 1

Reviewer: ```suggestion
      ./chip-tool doorlock set-year-day-schedule 1 1 60 120 1 1
```

> #. Set the example ``holiday`` schedule using the following command:

Reviewer: ```suggestion
#. Set the example ``Holiday`` schedule using the following command:
```

>    ./chip-tool doorlock set-holiday-schedule HolidayIndex LocalStartTime LocalEndTime OperatingMode destination-id endpoint-id

Reviewer: ```suggestion
   .. parsed-literal::
      :class: highlight
      
      ./chip-tool doorlock set-holiday-schedule *holiday-index* *localtime-start* *localtime-end* *operating-mode* *node-id* *endpoint-id*
```

>    Where:
> 

Reviewer: ```suggestion
```

>    * ``HolidayIndex`` is an index of the new schedule, starting from ``1``.

Reviewer: ```suggestion
   * *holiday-index* is the index of the new schedule, starting from ``1``.
```

   * ``LocalStartTime`` is the starting time in Epoch Time.
   * ``LocalEndTime`` is the ending time in Epoch Time.
   * ``OperatingMode`` is the operating mode described in the :ref:`matter_lock_scheduled_timed_access` section of this guide.
   * ``node-id`` is the device node ID.
   * ``endpoint-id`` is the Matter door lock endpoint, in this sample assigned to ``1``.

>    For example, to setup a ``holiday-schedule`` with type ``Vacation`` use the following command:

Reviewer: Expand this to explain all the values used as well.

Author: Done

   .. code-block:: console

      ./chip-tool doorlock set-holiday-schedule 1 60 120 1 1 1

> #. Read saved schedules using the following commands providing the arguments in a similar way as above:

Reviewer: ```suggestion
#. Read saved schedules using the following commands and providing the same arguments you used in the earlier steps:
```

   * *endpoint-id* is the Matter door lock endpoint, in this sample assigned to ``1``.

   For example, use the following command to set a ``Week-day`` schedule with index ``1`` for Tuesday, Thursday and Saturday to start at 7:30 AM and finish at 10:30 AM, dedicated for user with ID ``2``:

   .. code-block:: console

      ./chip-tool doorlock set-week-day-schedule 1 2 84 7 30 10 30 1 1

#. Set the example ``Year-day`` schedule using the following command:

   .. parsed-literal::
      :class: highlight

      ./chip-tool doorlock set-year-day-schedule *yearday-index* *user-index* *localtime-start* *localtime-end* *destination-id* *endpoint-id*

   * *yearday-index* is the index of the new schedule, starting from ``1``.
     The maximum value is defined by the :kconfig:option:`CONFIG_LOCK_MAX_YEARDAY_SCHEDULES_PER_USER` kconfig option.
   * *user-index* is the user index defined for the user created in the previous step.
   * *localtime-start* is the starting time in Epoch Time.
   * *localtime-end* is the ending time in Epoch Time.
   * *destination-id* is the device node ID.
   * *endpoint-id* is the Matter door lock endpoint, in this sample assigned to ``1``.

>    Both ``LocalStartTime`` and ``LocalEndTime`` are in seconds with the local time offset based on the local timezone and DST offset on the day represented by the value.

Reviewer: ```suggestion
   Both ``localtime-start`` and ``localtime-end`` are in seconds with the local time offset based on the local timezone and DST offset on the day represented by the value.
```

   For example, use the following command to set a ``Year-day`` schedule with index ``1`` to start on Monday, May 27, 2024, at 7:00:00 AM GMT+02:00 DST and finish on Thursday, May 30, 2024, at 7:00:00 AM GMT+02:00 DST dedicated for user with ID ``2``::

   .. code-block:: console

      ./chip-tool doorlock set-year-day-schedule 1 2 1716786000 1717045200 1 1

#. Set the example ``Holiday`` schedule using the following command:

   .. parsed-literal::
      :class: highlight

      ./chip-tool doorlock set-holiday-schedule *holiday-index* *localtime-start* *localtime-end* *operating-mode* *destination-id* *endpoint-id*

   * *holiday-index* is the index of the new schedule, starting from ``1``.
   * *localtime-start* is the starting time in Epoch Time.

>    * *operating-mode* is the ending time in Epoch Time.

Reviewer: ```suggestion
   * *localtime-end* is the ending time in Epoch Time.
```

   * *localtime-end* is the ending time in Epoch Time.
   * *operating-mode* is the operating mode described in the :ref:`matter_lock_scheduled_timed_access` section of this guide.
   * *destination-id* is the device node ID.
   * *endpoint-id* is the Matter door lock endpoint, in this sample assigned to ``1``.

   For example, use the following command to setup a ``Holiday`` schedule with the operating mode ``Vacation`` to start on Monday, May 27, 2024, at 7:00:00 AM GMT+02:00 DST and finish on Thursday, May 30, 2024, at 7:00:00 AM GMT+02:00 DST:

   .. parsed-literal::
      :class: highlight

      ./chip-tool doorlock set-holiday-schedule 1 1716786000 1717045200 1 1 1

#. Read saved schedules using the following commands and providing the same arguments you used in the earlier steps:

   .. parsed-literal::
      :class: highlight

      ./chip-tool doorlock get-week-day-schedule *weekday-index* *user-index* *destination-id* *endpoint-id*

   .. parsed-literal::
      :class: highlight

      ./chip-tool doorlock get-year-day-schedule *yearday-index* *user-index* *destination-id* *endpoint-id*

   .. parsed-literal::
      :class: highlight

>       ./chip-tool doorlock get-holiday-schedule *holiday-index* *destination-id* *endpoint-id*

Reviewer 2: Should we add some commends to actually test if the access works? Like opening the bolt lock in defined time slot?
